common/fs/src/cache/watch.rs

```diff
@@ -55,82 +72,210 @@ impl Watcher {
         self.inotify.rm_watch(wd)
     }
 
-    pub fn read_events(&mut self, buffer: &mut [u8]) -> io::Result<Vec<WatchEvent>> {
-        let mut events = Vec::new();
-        for raw_event in self.inotify.read_events(buffer)? {
-            if raw_event.mask.contains(EventMask::MOVED_FROM) {
-                let mut found_match = false;
-                for event in events.iter_mut() {
-                    match event {
-                        WatchEvent::MovedTo { wd, name, cookie } => {
-                            if *cookie == raw_event.cookie {
-                                *event = WatchEvent::Move {
-                                    from_wd: raw_event.wd.clone(),
-                                    from_name: raw_event.name.unwrap().to_os_string(),
-                                    to_wd: wd.clone(),
-                                    to_name: name.clone(),
-                                };
-                                found_match = true;
-                                break;
-                            }
-                        }
-                        _ => continue,
-                    };
-                }
+    pub fn event_stream<'a>(
+        &mut self,
+        buffer: &'a mut [u8],
+    ) -> std::io::Result<WatchEventStream<'a>> {
+        Ok(WatchEventStream {
+            event_stream: self.inotify.event_stream(buffer)?,
+        })
```

Comment:
Why return the wrapper `WatchEventStream` around the stream? Why not just return the stream?

Reply:
Mostly it was just for the convenience of having a type to impl against rather that whatever monstrous thing into_stream's concrete return type resolves too. A tangential benefit is that by wrapping it up it allows us to return the stream and strart processing it with WatchEventStream::into_stream in separate steps.

I could just return the impl Stream here, but it would be a bit less flexible, do you have a strong feeling either way?

Reply:
Whatever we do we should be consistent. I don't have an issue with doing things as they are now. In my implementation of converting a Journald instance into a stream I'll try and implement something similar for the same minor benefits.